### ISSUE TITLE
BGP-LS session throws TypeError exception for messages with ISIS wide igp-metric TLV (1095)

### ISSUE THREAD

**Issue by dadamfon** [Exa-Networks/exabgp#1071]
**Bug description**

Unexpected **TypeError: int() argument must be a string, a bytes-like object or a number, not 'list'** while decoding messages with igp-metric drop the connection.

**To Reproduce**
1) Install exabgp 4.2.17 (also tested in master branch with the same result)

2) Setup exabgp.conf:

```
neighbor 192.168.252.208 {
        description bgp-ls-RR;
        local-address 80.58.171.201;
        local-as 64531;
        peer-as 64531;
        hold-time 90;
        family {
                  bgp-ls  bgp-ls;
        }
        capability {
               asn4 enable;
               route-refresh enable;
               graceful-restart 120;
       }
}
```
3) run exabgp service

```
$ /usr/local/bin/exabgp /root/exabgp-master/exabgp.conf -d -e /usr/local/bin/etc/exabgp/exabgp.env
```

**Expected behavior**
Expecting compatibility decoding messages with "igp-metric" to not drop the connection.

**Environment:**

 - OS: Red Hat Enterprise Linux 8.3 (Ootpa)
 - Version exabgp 4.2.17 (tested also in master branch)

**Additional context**

We are using exabgp to establish a bgp-ls session with a route reflector. After running exabgp, the session gets established but a few seconds later this session drops because of the error: the speaker is not able to decode a specific type of bgp updates messages. Checking those update messages, and seeing the error is related to the igp-metric we found that this messages always have the ISIS wide igp-metric TLV (1095).

Trackeback log:
```
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      | received complete TCP payload (  19) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 00B6 02
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      | received complete TCP payload ( 163) 0000 009F 900E 0062 4004 4704 C0A8 FCA3 0000 0200 5502 0000 0000 0000 004E 0100 001A 0200 0004 0000 FC13 0201 0004 0000 00B2 0203 0006                            
																															   1921 6825 2161 0101 001A 0200 0004 0000 FC13 0201 0004 0000 00B2 0203 0006 1921 6825 1125 0103 0004 C0A8 C307 0104 0004 C0A8 C306 4001
																															   0100 4002 0040 0504 0000 0064 8009 04C0 A8FC A380 0A04 0C04 0101 801D 1A04 4400 0400 0013 8804 4700 0300 1388 044B 0007 3000 0000 0034 B4
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      | << message of type UPDATE
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | parser          | parsing UPDATE ( 163) 0000 009F 900E 0062 4004 4704 C0A8 FCA3 0000 0200 5502 0000 0000 0000 004E 0100 001A 0200 0004 0000 FC13 0201 0004 0000 00B2 0203 0006 1921 6825 2161
																												0101 001A 0200 0004 0000 FC13 0201 0004 0000 00B2 0203 0006 1921 6825 1125 0103 0004 C0A8 C307 0104 0004 C0A8 C306 4001 0100 4002 0040 0504 0000 0064
																												8009 04C0 A8FC A380 0A04 0C04 0101 801D 1A04 4400 0400 0013 8804 4700 0300 1388 044B 0007 3000 0000 0034 B4
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | routes          | withdrawn NLRI none
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | parser          | attribute mp-reach-nlri      flag 0x90 type 0x0e len 0x62 payload 4004 4704 C0A8 FCA3 0000 0200 5502 0000 0000 0000 004E 0100 001A 0200 0004 0000 FC13 0201 0004 0000 00B2
																																							0203 0006 1921 6825 2161 0101 001A 0200 0004 0000 FC13 0201 0004 0000 00B2 0203 0006 1921 6825 1125 0103
																																							0004 C0A8 C307 0104 0004 C0A8 C306
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | parser          | NLRI      bgp-ls bgp-ls      without path-information     payload 0002 0055 0200 0000 0000 0000 4E01 0000 1A02 0000 0400 00FC 1302 0100 0400 0000 B202 0300 0619 2168 2521
																																							6101 0100 1A02 0000 0400 00FC 1302 0100 0400 0000 B202 0300 0619 2168 2511 2501 0300 04C0 A8C3 0701 0400 
																																							04C0 A8C3 06
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | parser          | attribute origin             flag 0x40 type 0x01 len 0x01 payload 00
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | parser          | attribute as-path            flag 0x40 type 0x02 len 0x00
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | parser          | attribute local-preference   flag 0x40 type 0x05 len 0x04 payload 0000 0064
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | parser          | attribute originator-id      flag 0x80 type 0x09 len 0x04 payload C0A8 FCA3
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | parser          | attribute cluster-list       flag 0x80 type 0x0a len 0x04 payload 0C04 0101
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | parser          | attribute bgp-ls             flag 0x80 type 0x1d len 0x1a payload 0444 0004 0000 1388 0447 0003 0013 8804 4B00 0730 0000 0000 34B4
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | routes          | announced NLRI none
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      | could not decode message "2"
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      | int() argument must be a string, a bytes-like object or a number, not 'list'
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      | Traceback (most recent call last):
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |   File "/usr/local/lib/python3.6/site-packages/exabgp-4.2.17-py3.6.egg/exabgp/reactor/protocol.py", line 240, in read_message
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |     message = Message.unpack(msg_id, body, Direction.IN, self.negotiated)
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |   File "/usr/local/lib/python3.6/site-packages/exabgp-4.2.17-py3.6.egg/exabgp/bgp/message/message.py", line 180, in unpack
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |     return cls.klass(message).unpack_message(data, direction, negotiated)
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |   File "/usr/local/lib/python3.6/site-packages/exabgp-4.2.17-py3.6.egg/exabgp/bgp/message/update/__init__.py", line 295, in unpack_message
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |     logger.debug(LazyFormat('decoded UPDATE', '', parsed), 'parser')
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |   File "/usr/local/lib/python3.6/site-packages/exabgp-4.2.17-py3.6.egg/exabgp/logger.py", line 355, in debug
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |     self._report(message, source, level)
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |   File "/usr/local/lib/python3.6/site-packages/exabgp-4.2.17-py3.6.egg/exabgp/logger.py", line 347, in _report
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |     for line in message.split('\n'):
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |   File "/usr/local/lib/python3.6/site-packages/exabgp-4.2.17-py3.6.egg/exabgp/logger.py", line 53, in split
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |     return str(self).split(char)
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |   File "/usr/local/lib/python3.6/site-packages/exabgp-4.2.17-py3.6.egg/exabgp/logger.py", line 56, in __str__
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |     formated = self.formater(self.message)
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |   File "/usr/local/lib/python3.6/site-packages/exabgp-4.2.17-py3.6.egg/exabgp/bgp/message/update/__init__.py", line 293, in parsed
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |     return 'json %s' % Response.JSON(json_version).update(negotiated.neighbor, 'in', update, None, '', '')
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |   File "/usr/local/lib/python3.6/site-packages/exabgp-4.2.17-py3.6.egg/exabgp/reactor/api/response/json.py", line 321, in update
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |     message = self._update(update)
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |   File "/usr/local/lib/python3.6/site-packages/exabgp-4.2.17-py3.6.egg/exabgp/reactor/api/response/json.py", line 312, in _update
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |     attributes = '' if not update.attributes else '"attribute": { %s }' % update.attributes.json()
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |   File "/usr/local/lib/python3.6/site-packages/exabgp-4.2.17-py3.6.egg/exabgp/bgp/message/update/attribute/attributes.py", line 272, in json
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |     self._json = ', '.join(self._generate_json())
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |   File "/usr/local/lib/python3.6/site-packages/exabgp-4.2.17-py3.6.egg/exabgp/bgp/message/update/attribute/attributes.py", line 174, in _generate_json
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |     yield '"%s": %s' % (name, presentation % attribute.json())
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |   File "/usr/local/lib/python3.6/site-packages/exabgp-4.2.17-py3.6.egg/exabgp/bgp/message/update/attribute/bgpls/linkstate.py", line 70, in json
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |     content = ', '.join(d.json() for d in self.ls_attrs)
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |   File "/usr/local/lib/python3.6/site-packages/exabgp-4.2.17-py3.6.egg/exabgp/bgp/message/update/attribute/bgpls/linkstate.py", line 70, in <genexpr>
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |     content = ', '.join(d.json() for d in self.ls_attrs)
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |   File "/usr/local/lib/python3.6/site-packages/exabgp-4.2.17-py3.6.egg/exabgp/bgp/message/update/attribute/bgpls/link/igpmetric.py", line 62, in json
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |     return '"igp-metric": %d' % int(self.igpmetric)
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      | TypeError: int() argument must be a string, a bytes-like object or a number, not 'list'
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      |
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      | sending TCP payload (  62) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 003E 0301 0063 616E 206E 6F74 2064 6563 6F64 6520 7570 6461 7465 206D 6573 7361 6765 206F 6620 7479 7065 2022 3222
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      | >> NOTIFICATION (1,0,"can not decode update message of type "2"")
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      | peer reset, message [notification sent (1,0)] error[Message header error / Unspecific / can not decode update message of type "2"]
Feb 15 09:40:40 localhost.localdomain exabgp[5912]: 09:40:40 | 5912   | outgoing-6      | outgoing-6 80.58.171.201-192.168.252.208, closing connection
```

**Comment by thomas-mangin** [comment]
Thank you for assigning this issue to me. I shall take this a request for consulting and will invoice you for the time I spend looking into the issue.

Obviously, I am not serious but I am sure you can understand the feeling I have when people I do not know assign work to me on Github, like if they were my boss, oh, and I have did not have a boss for the last 20 years as I founded my own business.
TLDR, I need to find how to disable this "feature".

More seriously, I did not author that code but I did refactor it in master, fixing many issues with it. I do not intend on back porting the work to 4.2 - too much work.

If you had tested master as the README suggests before reporting bugs, you may have been happily surprised there .. but it is buggy too 🤦 

```sh
> ./sbin/exabgp decode -c etc/exabgp/conf-bgpls.conf "FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 00B6 02 0000 009F 900E 0062 4004 4704 C0A8 FCA3 0000 0200 5502 0000 0000 0000 004E 0100 001A 0200 0004 0000 FC13 0201 0004 0000 00B2 0203 0006 1921 6825 2161 0101 001A 0200 0004 0000 FC13 0201 0004 0000 00B2 0203 0006 1921 6825 1125 0103 0004 C0A8 C307 0104 0004 C0A8 C306 4001 0100 4002 0040 0504 0000 0064 8009 04C0 A8FC A380 0A04 0C04 0101 801D 1A04 4400 0400 0013 8804 4700 0300 1388 044B 0007 3000 0000 0034 B4"
configuration performing reload of exabgp master-350d430837cd0ef7c71c0b88f5ebcb143f2907e0
reactor       loaded new configuration successfully
parser        decoded update 1 { "ls-nlri-type": "bgpls-link", "l3-routing-topology": 78, "protocol-id": 2, "local-node-descriptors": { "autonomous-system": 64531, "bgp-ls-identifier": "178", "router-id": "192168252161" }, "remote-node-descriptors": { "autonomous-system": 64531, "bgp-ls-identifier": "178", "router-id": "192168251125" }, "interface-address": { "interface-address": "192.168.195.7" }, "neighbor-address": { "neighbor-address": "192.168.195.6" } } origin igp local-preference 100 originator-id 192.168.252.163 cluster-list [ 12.4.1.1 ] bgp-ls TE Default Metric: 5000, IGP Metric: 5000, adj_flags: {'F': 0, 'B': 0, 'V': 1, 'L': 1, 'S': 0, 'P': 0, 'RSV': 0}, sids: [13492], undecoded_sid []
parser        update json { "exabgp": "4.0.1", "time": 1648590778.626667, "host" : "MacBook-Pro-2.local", "pid" : 34381, "ppid" : 74362, "counter": 1, "type": "update", "neighbor": {     "address": { "local": "127.0.0.1", "peer": "127.0.0.1" },     "asn": { "local": 1, "peer": 1 }     , "direction": "in", "message": { "update": { "attribute": { "origin": "igp", "local-preference": 100, "originator-id": "192.168.252.163", "cluster-list": [ "12.4.1.1" ], "bgp-ls": { "te-metric": 5000, "igp-metric": 5000, {"sr-adj-flags": {"F": 0, "B": 0, "V": 1, "L": 1, "S": 0, "P": 0, "RSV": 0}, "sids": [13492], "undecoded-sids": [], "sr-adj-weight": 0} } }, "announce": { "bgp-ls bgp-ls": { "192.168.252.163": [ { "ls-nlri-type": "bgpls-link", "l3-routing-topology": 78, "protocol-id": 2, "local-node-descriptors": { "autonomous-system": 64531, "bgp-ls-identifier": "178", "router-id": "192168252161" }, "remote-node-descriptors": { "autonomous-system": 64531, "bgp-ls-identifier": "178", "router-id": "192168251125" }, "interface-address": { "interface-address": "192.168.195.7" }, "neighbor-address": { "neighbor-address": "192.168.195.6" } } ] } } } } } }
```

So I will fix it in master .. free of charge 😉

**Comment by thomas-mangin** [comment]
Seems to be working now:
```sh
> ./sbin/exabgp decode -c etc/exabgp/conf-bgpls.conf "FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 00B6 02 0000 009F 900E 0062 4004 4704 C0A8 FCA3 0000 0200 5502 0000 0000 0000 004E 0100 001A 0200 0004 0000 FC13 0201 0004 0000 00B2 0203 0006 1921 6825 2161 0101 001A 0200 0004 0000 FC13 0201 0004 0000 00B2 0203 0006 1921 6825 1125 0103 0004 C0A8 C307 0104 0004 C0A8 C306 4001 0100 4002 0040 0504 0000 0064 8009 04C0 A8FC A380 0A04 0C04 0101 801D 1A04 4400 0400 0013 8804 4700 0300 1388 044B 0007 3000 0000 0034 B4" 2>&1 | tail -n 1 | cut -b29- | jsonpp
```
```json
{
  "exabgp": "5.0.0",
  "time": 1648592185.66445,
  "host": "MacBook-Pro-2.local",
  "pid": 36608,
  "ppid": 74362,
  "counter": 1,
  "type": "update",
  "neighbor": {
    "address": {
      "local": "127.0.0.1",
      "peer": "127.0.0.1"
    },
    "asn": {
      "local": 1,
      "peer": 1
    },
    "direction": "in",
    "message": {
      "update": {
        "attribute": {
          "origin": "igp",
          "local-preference": 100,
          "originator-id": "192.168.252.163",
          "cluster-list": [
            "12.4.1.1"
          ],
          "bgp-ls": {
            "te-metric": 5000,
            "igp-metric": 5000,
            "sr-adj": {
              "flags": {
                "F": 0,
                "B": 0,
                "V": 1,
                "L": 1,
                "S": 0,
                "P": 0,
                "RSV": 0
              },
              "sids": [
                13492
              ],
              "weight": 0,
              "undecoded-sids": []
            }
          }
        },
        "announce": {
          "bgp-ls bgp-ls": {
            "192.168.252.163": [
              {
                "ls-nlri-type": "bgpls-link",
                "l3-routing-topology": 78,
                "protocol-id": 2,
                "local-node-descriptors": {
                  "autonomous-system": 64531,
                  "bgp-ls-identifier": "178",
                  "router-id": "192168252161"
                },
                "remote-node-descriptors": {
                  "autonomous-system": 64531,
                  "bgp-ls-identifier": "178",
                  "router-id": "192168251125"
                },
                "interface-address": {
                  "interface-address": "192.168.195.7"
                },
                "neighbor-address": {
                  "neighbor-address": "192.168.195.6"
                }
              }
            ]
          }
        }
      }
    }
  }
}
```

**Comment by thomas-mangin** [comment]
So thank you for providing the information I needed to fix the issue, hopefully you can get back to what you wanted to do with BGP 👋

**Comment by dadamfon** [comment]
Amazing! Thank you @thomas-mangin for the quick response and fix! We will test it soon.

**Comment by Patricia08** [comment]
Hello!

Sorry for the late response but we had some issues on our environment so we couldn´t test this until now.

We have tested with lastest master version. We don´t get the last error anymore but there is a new one, "TypeError: 'int' object is not subscriptable":

abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9      received TCP payload (  19) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 00AF 02
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9      received TCP payload ( 156) 0000 0098 4001 0100 4002 0040 0504 0000 0064 8009 04C0 A8FC B280 0A04 0C04 0101 801D 1301 0200 0800 0001 7200 0001 BB04 4700 0300 1388 900E 0062 4004 4704 C0A8 FCB2 0000 0200 5502 0000 0000 0000 0002 0100 001A 0200 0004 0000 0D18 0201 0004 0000 00B2 0203 0006 1921 6825 2240 0101 001A 0200 0004 0000 0D18 0201 0004 0000 00B2 0203 0006 1921 6825 2162 0103 0004 C0A8 C754 0104 0004 C0A8 C755
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9      << message of type UPDATE
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  routes          withdrawn NLRI none
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9      could not decode message "2"
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9      'int' object is not subscriptable
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9      Traceback (most recent call last):
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9        File "/usr/local/lib/python3.7/site-packages/exabgp-master-py3.7.egg/exabgp/reactor/protocol.py", line 226, in read_message
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9          message = Message.unpack(msg_id, body, Direction.IN, self.negotiated)
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9        File "/usr/local/lib/python3.7/site-packages/exabgp-master-py3.7.egg/exabgp/bgp/message/message.py", line 177, in unpack
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9          return cls.klass(message).unpack_message(data, direction, negotiated)
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9        File "/usr/local/lib/python3.7/site-packages/exabgp-master-py3.7.egg/exabgp/bgp/message/update/__init__.py", line 270, in unpack_message
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9          attributes = Attributes.unpack(_attributes, direction, negotiated)
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9        File "/usr/local/lib/python3.7/site-packages/exabgp-master-py3.7.egg/exabgp/bgp/message/update/attribute/attributes.py", line 290, in unpack
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9          attributes = cls().parse(data, direction, negotiated)
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9        File "/usr/local/lib/python3.7/site-packages/exabgp-master-py3.7.egg/exabgp/bgp/message/update/attribute/attributes.py", line 385, in parse
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9          return self.parse(left, direction, negotiated)
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9        File "/usr/local/lib/python3.7/site-packages/exabgp-master-py3.7.egg/exabgp/bgp/message/update/attribute/attributes.py", line 385, in parse
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9          return self.parse(left, direction, negotiated)
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9        File "/usr/local/lib/python3.7/site-packages/exabgp-master-py3.7.egg/exabgp/bgp/message/update/attribute/attributes.py", line 385, in parse
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9          return self.parse(left, direction, negotiated)
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9        [Previous line repeated 2 more times]
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9        File "/usr/local/lib/python3.7/site-packages/exabgp-master-py3.7.egg/exabgp/bgp/message/update/attribute/attributes.py", line 371, in parse
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9          decoded = Attribute.unpack(aid, flag, attribute, direction, negotiated)
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9        File "/usr/local/lib/python3.7/site-packages/exabgp-master-py3.7.egg/exabgp/bgp/message/update/attribute/attribute.py", line 278, in unpack
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9          instance = cls.klass(attribute_id, flag).unpack(data, direction, negotiated)
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9        File "/usr/local/lib/python3.7/site-packages/exabgp-master-py3.7.egg/exabgp/bgp/message/update/attribute/bgpls/linkstate.py", line 60, in unpack
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9          klass = cls.klass(scode).unpack(data[4 : length + 4], length)
**abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9        File "/usr/local/lib/python3.7/site-packages/exabgp-master-py3.7.egg/exabgp/bgp/message/update/attribute/bgpls/linkstate.py", line 115, in unpack
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9          return cls(scode, binascii.b2a_uu(data[:]))
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9      TypeError: 'int' object is not subscriptable**
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9      sending TCP payload (  62) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 003E 0301 0063 616E 206E 6F74 2064 6563 6F64 6520 7570 6461 7465 206D 6573 7361 6765 206F 6620 7479 7065 2022 3222
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9      >> NOTIFICATION (1,0,"can not decode update message of type "2"")
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9      peer reset, message [notification sent (1,0)] error[Message header error / Unspecific / can not decode update message of type "2"]
abr 18 13:38:00 localhost.localdomain exabgp[51630]: 13:38:00 51630  outgoing-9      outgoing-9 80.58.171.201-192.168.255.89, closing connection

This time we don´t know what is related this problem to, but we have captured the update message in case it is somehow helpful to detect the problem, it is in the word attached.

[update_msg_exabgp_error.docx](https://github.com/Exa-Networks/exabgp/files/8505523/update_msg_exabgp_error.docx)

Thank you so much for your help :)

**Comment by thomas-mangin** [comment]
Ok, thank you for the packet, I found why the decoding was failing and added it to the new decode tests. So should be working as expected now:

```sh
./sbin/exabgp decode -f 'bgp-ls bgp-ls' '0000 0098 4001 0100 4002 0040 0504 0000 0064 8009 04C0 A8FC B280 0A04 0C04 0101 801D 1301 0200 0800 0001 7200 0001 BB04 4700 0300 1388 900E 0062 4004 4704 C0A8 FCB2 0000 0200 5502 0000 0000 0000 0002 0100 001A 0200 0004 0000 0D18 0201 0004 0000 00B2 0203 0006 1921 6825 2240 0101 001A 0200 0004 0000 0D18 0201 0004 0000 00B2 0203 0006 1921 6825 2162 0103 0004 C0A8 C754 0104 0004 C0A8 C755'
configuration performing reload of exabgp master-7c116a9045768328d646471fc64f932e7f35e152
reactor       loaded new configuration successfully
parser        decoded update 1 { "ls-nlri-type": "bgpls-link", "l3-routing-topology": 2, "protocol-id": 2, "local-node-descriptors": { "autonomous-system": 3352, "bgp-ls-identifier": "178", "router-id": "192168252240" }, "remote-node-descriptors": { "autonomous-system": 3352, "bgp-ls-identifier": "178", "router-id": "192168252162" }, "interface-address": { "interface-address": "192.168.199.84" }, "neighbor-address": { "neighbor-address": "192.168.199.85" } } origin igp local-preference 100 originator-id 192.168.252.178 cluster-list [ 12.4.1.1 ] bgp-ls Attribute with code [ 258 ] not implemented, IGP Metric: 5000
parser        update json { "exabgp": "5.0.0", "time": 1650738183.056318, "host" : "MacBook-Pro-2.local", "pid" : 17989, "ppid" : 88681, "counter": 1, "type": "update", "neighbor": {     "address": { "local": "127.0.0.1", "peer": "127.0.0.1" },     "asn": { "local": 65533, "peer": 65533 }     , "direction": "in", "message": { "update": { "attribute": { "origin": "igp", "local-preference": 100, "originator-id": "192.168.252.178", "cluster-list": [ "12.4.1.1" ], "bgp-ls": { "generic-lsid-258": "0x28202020213C402020203B4C200A", "igp-metric": 5000 } }, "announce": { "bgp-ls bgp-ls": { "192.168.252.178": [ { "ls-nlri-type": "bgpls-link", "l3-routing-topology": 2, "protocol-id": 2, "local-node-descriptors": { "autonomous-system": 3352, "bgp-ls-identifier": "178", "router-id": "192168252240" }, "remote-node-descriptors": { "autonomous-system": 3352, "bgp-ls-identifier": "178", "router-id": "192168252162" }, "interface-address": { "interface-address": "192.168.199.84" }, "neighbor-address": { "neighbor-address": "192.168.199.85" } } ] } } } } } }
```
